### PR TITLE
got basic profile viewing working from answers page

### PR DESCRIPTION
**Pull request by spall03**

### CODE REVIEW

**Comment by spall03** [comment]
This commit features:
1) basic profile viewing from answers page when selecting an answer cell
2) answer cells show name and image of answering user
3) logout functionality from your own profile page (sort of)

there's a FIXME in 3), I figured out how to log out and get back to the login screen, but I can't seem to login again...

File: BlocQuery/BQProfileViewController.m [hunk 1]
```diff
@@ -154,21 +154,27 @@ - (void)viewDidLoad
...
+    //make profile editing buttons and leave them hidden by default
+    self.editUserImageButton = [UIButton buttonWithType:UIButtonTypeRoundedRect];
+    [self.editUserImageButton setTitle:@"Change Photo" forState:UIControlStateNormal];
+    [self.editUserImageButton addTarget:self action:@selector(editImageButtonPressed:) forControlEvents:UIControlEventTouchUpInside];
+    self.editUserImageButton.hidden = YES;
+    
+    self.editUserDescriptionButton = [UIButton buttonWithType:UIButtonTypeRoundedRect];
+    [self.editUserDescriptionButton setTitle:@"Edit Description" forState:UIControlStateNormal];
+    [self.editUserDescriptionButton addTarget:self action:@selector(editDescriptionButtonPressed:) forControlEvents:UIControlEventTouchUpInside];
+    self.editUserDescriptionButton.hidden = YES;
```
Comment (mhanlon) [inline BlocQuery/BQProfileViewController.m:165]: We also want to make sure our UITextView doesn't accept user input... at the moment the person can tap on the description and mess with the text on a person's profile.

File: BlocQuery/BQUser.m [hunk 2]
```diff
@@ -30,6 +30,31 @@ + (void)load
     
 }
 
+- (PFFile*)setDefaultProfileImage
```
Comment (mhanlon) [inline BlocQuery/BQUser.m]: I would rename this method:

```
+ (PFFile*)defaultProfileImage
```

As it is I'd expect this method to take a parameter (a `PFFile`) and set that image as the new default.

File: BlocQuery/BQAnswerTableViewController.m [hunk 3]
```diff
@@ -259,6 +262,21 @@ - (void)didPressAddAnswerButton:(id)sender
...
+    [tempQuery whereKey:@"username" equalTo:temp.userName];
+    
+    //load query into temporary user
+    BQUser *tempUser = [tempQuery findObjects][0];
+    
+    //create new profile screen for that user
+    BQProfileViewController *newProfileVC = [[BQProfileViewController alloc]initWithUser:tempUser];
+    
+    //go to that screen
+    [self.navigationController pushViewController:newProfileVC animated:YES];
```
Comment (mhanlon) [inline BlocQuery/BQAnswerTableViewController.m]: I know that this is temp and old data which is potentially garbage, but our current set of data in Parse gives us some good practice in defensive coding... at the moment, on one of our (null) users, we get the following crash:

```
2015-02-10 22:32:45.359 BlocQuery[9858:1275528] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 'Cannot do a comparison query for type: (null)'
*** First throw call stack:
(
    0   CoreFoundation                      0x00000001084cef35 __exceptionPreprocess + 165
    1   libobjc.A.dylib                     0x0000000108138bb7 objc_exception_throw + 45
    2   CoreFoundation                      0x00000001084cee6d +[NSException raise:format:] + 205
    3   BlocQuery                           0x00000001053c0a48 +[PFInternalUtils assertValidClassForQuery:] + 333
    4   BlocQuery                           0x00000001053912fe -[PFQuery whereKey:equalTo:] + 86
    5   BlocQuery                           0x000000010527be52 -[BQAnswerTableViewController tableView:didSelectRowAtIndexPath:] + 322
```

We'll want to verify that `temp` and `temp.userName != nil` in this scenario:

```
    [tempQuery whereKey:@"username" equalTo:temp.userName];
```

File: BlocQuery/BQUser.h [hunk 4]
```diff
@@ -18,6 +18,7 @@ static NSString* kBQDidPostNewAnswerToQuestion;
 @property (retain) NSString *userDescription;
 @property (retain) PFFile *userImage;
 
+- (PFFile*)setDefaultProfileImage;
```
Comment (mhanlon) [inline BlocQuery/BQUser.h]: See note in the implementation file.

**Comment by mhanlon** [comment]
Check out my comments... most of those should make sense, feel free to ping me for more info or clarification.

On the login/logout front, I'm going to check in a new branch, called `logout-login-fix-no-peeking`. See if you can figure it out without peeking, though, first.
So let me just drop a few hints... for our `PFLoginViewController` (and `PFSignupViewController`) to work  we needed to set ourselves as the delegate, in our prior code. How can we set something like that up again? Do we really want to push those controllers onto the stack on top of the `BQProfileViewController` we've navigated to? What does that mean for all the stuff we'd been viewing as (potentially) another user?
If we don't want to do it down in the `BQProfileViewController`, where would we do it? Is there a way to re-use what we did earlier, in the `AppDelegate.m`?
Think about those questions and see if you can't get it. And if you're still stumped or wrestling with it too much, take a look at the `logout-login-fix-no-peeking` branch, particularly the diff between that branch and this `viewing-other-profiles` branch.

**Comment by spall03** [comment]
Thanks Matt! I fixed the bugs you pointed out... I tried to figure out the logout fix myself for another 45 min but couldn't get it, so I peeked. I don't think I would've gotten to that solution within a reasonable timeframe, so thanks for setting that up! Also gave me some much-needed practice with Notifications...

**Comment by spall03** [comment]
so the last thing I'd like to do here is add a way to get to a user's profile off the question table view... I did this with the full cell selection method in the answer table view, but I forgot that I can't do the same thing on the question table view because that method is already calling the answer table! whoops.

any ideas? one thought would be to use the cell selection method to present a nav alert of some kind (offering the user a choice between "see profile" and "see answers"), but that seems like lousy UX to me...

**Comment by mhanlon** [comment]
> one thought would be to use the cell selection method to present a nav alert of some kind (offering the user a choice between "see profile" and "see answers"), but that seems like lousy UX to me...

100% agree. What you _could_ do is implement a custom cell class for this table view... if things get sticky, you could implement `hitTest:withEvent:` (See (https://developer.apple.com/library/ios/documentation/EventHandling/Conceptual/EventHandlingiPhoneOS/event_delivery_responder_chain/event_delivery_responder_chain.html) for more info) to see which UI element was touched and then respond accordingly...

**Comment by spall03** [comment]
New BQTableCellView class is plugged into both the Questions and the Answers tables, and links to user profiles and answers correctly. It's ugly but functional...

**Pull request merged**